// @ngrx-utils/store/fesm5/ngrx-utils-store.js
import { Directive, ViewContainerRef, TemplateRef, Input, NgModule, ContentChildren, ɵstringify, WrappedValue, ɵisPromise, ɵisObservable, Pipe, ChangeDetectorRef } from '@angular/core';
import { __spread, __values } from 'tslib';
import { NavigationEnd, Router, RouterLink, RouterLinkWithHref } from '@angular/router';
import { Observable } from 'rxjs';
import { takeUntil, pluck as pluck$1 } from 'rxjs/operators';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ngLet.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var NgLetContext = /** @class */ (function () {
    function NgLetContext() {
        this.$implicit = null;
        this.ngLet = null;
    }
    return NgLetContext;
}());
if (false) {
    /** @type {?} */
    NgLetContext.prototype.$implicit;
    /** @type {?} */
    NgLetContext.prototype.ngLet;
}
var NgLetDirective = /** @class */ (function () {
    function NgLetDirective(_vcr, _templateRef) {
        this._vcr = _vcr;
        this._templateRef = _templateRef;
        this._context = new NgLetContext();
    }
    Object.defineProperty(NgLetDirective.prototype, "ngLet", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._context.$implicit = this._context.ngLet = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NgLetDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._vcr.createEmbeddedView(this._templateRef, this._context);
    };
    NgLetDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngLet]',
                },] }
    ];
    /** @nocollapse */
    NgLetDirective.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: TemplateRef }
    ]; };
    NgLetDirective.propDecorators = {
        ngLet: [{ type: Input }]
    };
    return NgLetDirective;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgLetDirective.prototype._context;
    /**
     * @type {?}
     * @private
     */
    NgLetDirective.prototype._vcr;
    /**
     * @type {?}
     * @private
     */
    NgLetDirective.prototype._templateRef;
}
var NgLetModule = /** @class */ (function () {
    function NgLetModule() {
    }
    NgLetModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [NgLetDirective],
                    exports: [NgLetDirective],
                },] }
    ];
    return NgLetModule;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/operators/untilDestroy.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// create a symbol identify the observable I add to
// the component so it doesn't conflict with anything.
// I need this so I'm able to add the desired behaviour to the component.
/** @type {?} */
var destroy$ = Symbol('destroy$');
/**
 * An operator that takes until destroy it takes a components this a parameter
 * returns a pipeable RxJS operator.
 * @type {?}
 */
var untilDestroy = (/**
 * @template T
 * @param {?} component
 * @return {?}
 */
function (component) {
    /** @type {?} */
    var orignalDestroy = component.ngOnDestroy;
    if (orignalDestroy == null) {
        // Angular does not support dynamic added destroy methods
        // so make sure there is one.
        throw new Error('untilDestroy operator needs the component to have an ngOnDestroy method');
    }
    if (component[destroy$] === undefined) {
        // only hookup each component once.
        addDestroyObservableToComponent(component);
    }
    // pipe in the takeUntil destroy$ and return the source unaltered
    return takeUntil(component[destroy$]);
});
/**
 * \@internal
 * @param {?} component
 * @return {?}
 */
function addDestroyObservableToComponent(component) {
    component[destroy$] = new Observable((/**
     * @param {?} observer
     * @return {?}
     */
    function (observer) {
        // keep track of the original destroy function,
        // the user might do something in there
        /** @type {?} */
        var orignalDestroy = component.ngOnDestroy;
        // replace the ngOndestroy
        component.ngOnDestroy = (/**
         * @return {?}
         */
        function () {
            // fire off the destroy observable
            observer.next();
            // complete the observable
            observer.complete();
            // and at last, call the original destroy
            orignalDestroy.call(component);
        });
        // return cleanup function.
        return (/**
         * @param {?} _
         * @return {?}
         */
        function (_) { return (component[destroy$] = undefined); });
    }));
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/operators/pluck.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @template T, V
 * @param {...?} props
 * @return {?}
 */
function pluck() {
    var props = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        props[_i] = arguments[_i];
    }
    return pluck$1.apply(void 0, __spread(props));
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/operators/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/routerLinkMatch.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
function MatchExp() { }
/**
 * This directive will give you ability to add a class to the element
 * when router url match a regular expression.
 * The syntax is same with ngClass but replace the true/false expression
 * with your string based regexp (like the string you pass to new RegExp(''))
 *
 * \@example
 * Example: active-class will be added to a tag when router URL
 * contains this segment: products/12345
 *
 * ```html
 * <a routerLink="/products"
 *  [routerLinkMatch]="{
 *    "active-class": "products/\\d+"
 *  }"></a>
 * ```
 */
var RouterLinkMatch = /** @class */ (function () {
    function RouterLinkMatch(router, _renderer, _ngEl) {
        var _this = this;
        this.router = router;
        this._renderer = _renderer;
        this._ngEl = _ngEl;
        router.events.pipe(untilDestroy(this)).subscribe((/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            if (e instanceof NavigationEnd) {
                _this._curRoute = ((/** @type {?} */ (e))).urlAfterRedirects;
                _this._update();
            }
        }));
    }
    Object.defineProperty(RouterLinkMatch.prototype, "routerLinkMatch", {
        set: /**
         * @param {?} matchExp
         * @return {?}
         */
        function (matchExp) {
            if (matchExp && typeof matchExp === 'object') {
                this._matchExp = matchExp;
            }
            else {
                throw new TypeError("Unexpected type '" + typeof matchExp + "' of value for " +
                    "input of routerLinkMatch directive, expected 'object'");
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    RouterLinkMatch.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this._update();
    };
    /**
     * @return {?}
     */
    RouterLinkMatch.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.links.changes.pipe(untilDestroy(this)).subscribe((/**
         * @return {?}
         */
        function () { return _this._update(); }));
        this.linksWithHrefs.changes
            .pipe(untilDestroy(this))
            .subscribe((/**
         * @return {?}
         */
        function () { return _this._update(); }));
        this._update();
    };
    /**
     * @private
     * @return {?}
     */
    RouterLinkMatch.prototype._update = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.links || !this.linksWithHrefs || !this.router.navigated) {
            return;
        }
        /**
         * This a way of causing something to happen in the next micro-task / during a new round
         * of change detection.
         */
        Promise.resolve().then((/**
         * @return {?}
         */
        function () {
            var e_1, _a;
            /** @type {?} */
            var matchExp = _this._matchExp;
            try {
                for (var _b = __values(Object.keys(matchExp)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var classes = _c.value;
                    if (matchExp[classes] && typeof matchExp[classes] === 'string') {
                        /** @type {?} */
                        var regexp = new RegExp(matchExp[classes]);
                        if (_this._curRoute.match(regexp)) {
                            _this._toggleClass(classes, true);
                        }
                        else {
                            _this._toggleClass(classes, false);
                        }
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }));
    };
    /**
     * @private
     * @param {?} classes
     * @param {?} enabled
     * @return {?}
     */
    RouterLinkMatch.prototype._toggleClass = /**
     * @private
     * @param {?} classes
     * @param {?} enabled
     * @return {?}
     */
    function (classes, enabled) {
        var _this = this;
        classes
            .split(/\s+/g)
            .filter((/**
         * @param {?} cls
         * @return {?}
         */
        function (cls) { return !!cls; }))
            .forEach((/**
         * @param {?} cls
         * @return {?}
         */
        function (cls) {
            if (enabled) {
                _this._renderer.addClass(_this._ngEl.nativeElement, cls);
            }
            else {
                _this._renderer.removeClass(_this._ngEl.nativeElement, cls);
            }
        }));
    };
    /**
     * @return {?}
     */
    RouterLinkMatch.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () { };
    RouterLinkMatch.decorators = [
        { type: Directive, args: [{
                    selector: '[routerLinkMatch]',
                },] }
    ];
    /** @nocollapse */
    RouterLinkMatch.ctorParameters = function () { return [
        { type: Router },
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    RouterLinkMatch.propDecorators = {
        links: [{ type: ContentChildren, args: [RouterLink, { descendants: true },] }],
        linksWithHrefs: [{ type: ContentChildren, args: [RouterLinkWithHref, { descendants: true },] }],
        routerLinkMatch: [{ type: Input, args: ['routerLinkMatch',] }]
    };
    return RouterLinkMatch;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    RouterLinkMatch.prototype._curRoute;
    /**
     * @type {?}
     * @private
     */
    RouterLinkMatch.prototype._matchExp;
    /** @type {?} */
    RouterLinkMatch.prototype.links;
    /** @type {?} */
    RouterLinkMatch.prototype.linksWithHrefs;
    /**
     * @type {?}
     * @private
     */
    RouterLinkMatch.prototype.router;
    /**
     * @type {?}
     * @private
     */
    RouterLinkMatch.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    RouterLinkMatch.prototype._ngEl;
}
var RouterLinkMatchModule = /** @class */ (function () {
    function RouterLinkMatchModule() {
    }
    RouterLinkMatchModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [RouterLinkMatch],
                    exports: [RouterLinkMatch],
                },] }
    ];
    return RouterLinkMatchModule;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/pipes/push.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} type
 * @param {?} value
 * @return {?}
 */
function invalidPipeArgumentError(type, value) {
    return Error("InvalidPipeArgument: '" + value + "' for pipe '" + ɵstringify(type) + "'");
}
/**
 * @record
 */
function SubscriptionStrategy() { }
if (false) {
    /**
     * @param {?} async
     * @param {?} updateLatestValue
     * @return {?}
     */
    SubscriptionStrategy.prototype.createSubscription = function (async, updateLatestValue) { };
    /**
     * @param {?} subscription
     * @return {?}
     */
    SubscriptionStrategy.prototype.dispose = function (subscription) { };
}
var ObservableStrategy = /** @class */ (function () {
    function ObservableStrategy() {
    }
    /**
     * @param {?} async
     * @param {?} updateLatestValue
     * @return {?}
     */
    ObservableStrategy.prototype.createSubscription = /**
     * @param {?} async
     * @param {?} updateLatestValue
     * @return {?}
     */
    function (async, updateLatestValue) {
        return async.subscribe({
            next: updateLatestValue,
            error: (/**
             * @param {?} e
             * @return {?}
             */
            function (e) {
                throw e;
            }),
        });
    };
    /**
     * @param {?} subscription
     * @return {?}
     */
    ObservableStrategy.prototype.dispose = /**
     * @param {?} subscription
     * @return {?}
     */
    function (subscription) {
        subscription.unsubscribe();
    };
    return ObservableStrategy;
}());
var PromiseStrategy = /** @class */ (function () {
    function PromiseStrategy() {
    }
    /**
     * @param {?} async
     * @param {?} updateLatestValue
     * @return {?}
     */
    PromiseStrategy.prototype.createSubscription = /**
     * @param {?} async
     * @param {?} updateLatestValue
     * @return {?}
     */
    function (async, updateLatestValue) {
        return async.then(updateLatestValue, (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            throw e;
        }));
    };
    /**
     * @param {?} subscription
     * @return {?}
     */
    PromiseStrategy.prototype.dispose = /**
     * @param {?} subscription
     * @return {?}
     */
    function (subscription) { };
    return PromiseStrategy;
}());
/** @type {?} */
var _promiseStrategy = new PromiseStrategy();
/** @type {?} */
var _observableStrategy = new ObservableStrategy();
/**
 * \@ngModule PushPipeModule
 * \@description
 *
 * Unwraps a value from an asynchronous primitive.
 *
 * The `push` pipe subscribes to an `Observable` or `Promise` and returns the latest value it has
 * emitted. When a new value is emitted, the `push` pipe will run change detection and it works
 * even when `zone` has been disabled. When the component gets destroyed,
 * the `push` pipe unsubscribes automatically to avoid potential memory leaks.
 *
 */
var PushPipe = /** @class */ (function () {
    function PushPipe(_ref) {
        this._ref = _ref;
        this._latestValue = null;
        this._latestReturnedValue = null;
        this._subscription = null;
        this._obj = null;
        this._strategy = (/** @type {?} */ (null));
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    PushPipe.prototype.transform = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        if (this._obj === null) {
            if (obj != null) {
                this._subscribe(obj);
            }
            this._latestReturnedValue = this._latestValue;
            return this._latestValue;
        }
        if (obj !== this._obj) {
            this._dispose();
            return this.transform((/** @type {?} */ (obj)));
        }
        if (this._latestValue === this._latestReturnedValue) {
            return this._latestReturnedValue;
        }
        this._latestReturnedValue = this._latestValue;
        return WrappedValue.wrap(this._latestValue);
    };
    /**
     * @return {?}
     */
    PushPipe.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._subscription !== null) {
            this._dispose();
        }
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    PushPipe.prototype._subscribe = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        this._obj = obj;
        this._strategy = this._selectStrategy(obj);
        this._subscription = this._strategy.createSubscription(obj, (/**
         * @param {?} value
         * @return {?}
         */
        function (value) { return _this._updateLatestValue(obj, value); }));
    };
    /**
     * @private
     * @return {?}
     */
    PushPipe.prototype._dispose = /**
     * @private
     * @return {?}
     */
    function () {
        this._strategy.dispose((/** @type {?} */ (this._subscription)));
        this._latestValue = null;
        this._latestReturnedValue = null;
        this._subscription = null;
        this._obj = null;
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    PushPipe.prototype._selectStrategy = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        if (ɵisPromise(obj)) {
            return _promiseStrategy;
        }
        if (ɵisObservable(obj)) {
            return _observableStrategy;
        }
        throw invalidPipeArgumentError(PushPipe, obj);
    };
    /**
     * @private
     * @param {?} async
     * @param {?} value
     * @return {?}
     */
    PushPipe.prototype._updateLatestValue = /**
     * @private
     * @param {?} async
     * @param {?} value
     * @return {?}
     */
    function (async, value) {
        if (async === this._obj) {
            this._latestValue = value;
            this._ref.detectChanges();
        }
    };
    PushPipe.decorators = [
        { type: Pipe, args: [{ name: 'push', pure: false },] }
    ];
    /** @nocollapse */
    PushPipe.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    return PushPipe;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._latestValue;
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._latestReturnedValue;
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._subscription;
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._obj;
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._strategy;
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._ref;
}
var PushPipeModule = /** @class */ (function () {
    function PushPipeModule() {
    }
    PushPipeModule.decorators = [
        { type: NgModule, args: [{
                    exports: [PushPipe],
                    declarations: [PushPipe],
                },] }
    ];
    return PushPipeModule;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/pipes/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/store.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: ngrx-utils-store.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { NgLetContext, NgLetDirective, NgLetModule, PushPipe, PushPipeModule, RouterLinkMatch, RouterLinkMatchModule, invalidPipeArgumentError, pluck, untilDestroy, destroy$ as ɵdestroy$ };
