// @ngrx-utils/store/esm5/lib/pipes/push.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pipes/push.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, NgModule, Pipe, WrappedValue, ɵisObservable as isObservable, ɵisPromise as isPromise, } from '@angular/core';
import { ɵstringify as stringify } from '@angular/core';
/**
 * @param {?} type
 * @param {?} value
 * @return {?}
 */
export function invalidPipeArgumentError(type, value) {
    return Error("InvalidPipeArgument: '" + value + "' for pipe '" + stringify(type) + "'");
}
/**
 * @record
 */
function SubscriptionStrategy() { }
if (false) {
    /**
     * @param {?} async
     * @param {?} updateLatestValue
     * @return {?}
     */
    SubscriptionStrategy.prototype.createSubscription = function (async, updateLatestValue) { };
    /**
     * @param {?} subscription
     * @return {?}
     */
    SubscriptionStrategy.prototype.dispose = function (subscription) { };
}
var ObservableStrategy = /** @class */ (function () {
    function ObservableStrategy() {
    }
    /**
     * @param {?} async
     * @param {?} updateLatestValue
     * @return {?}
     */
    ObservableStrategy.prototype.createSubscription = /**
     * @param {?} async
     * @param {?} updateLatestValue
     * @return {?}
     */
    function (async, updateLatestValue) {
        return async.subscribe({
            next: updateLatestValue,
            error: (/**
             * @param {?} e
             * @return {?}
             */
            function (e) {
                throw e;
            }),
        });
    };
    /**
     * @param {?} subscription
     * @return {?}
     */
    ObservableStrategy.prototype.dispose = /**
     * @param {?} subscription
     * @return {?}
     */
    function (subscription) {
        subscription.unsubscribe();
    };
    return ObservableStrategy;
}());
var PromiseStrategy = /** @class */ (function () {
    function PromiseStrategy() {
    }
    /**
     * @param {?} async
     * @param {?} updateLatestValue
     * @return {?}
     */
    PromiseStrategy.prototype.createSubscription = /**
     * @param {?} async
     * @param {?} updateLatestValue
     * @return {?}
     */
    function (async, updateLatestValue) {
        return async.then(updateLatestValue, (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            throw e;
        }));
    };
    /**
     * @param {?} subscription
     * @return {?}
     */
    PromiseStrategy.prototype.dispose = /**
     * @param {?} subscription
     * @return {?}
     */
    function (subscription) { };
    return PromiseStrategy;
}());
/** @type {?} */
var _promiseStrategy = new PromiseStrategy();
/** @type {?} */
var _observableStrategy = new ObservableStrategy();
/**
 * \@ngModule PushPipeModule
 * \@description
 *
 * Unwraps a value from an asynchronous primitive.
 *
 * The `push` pipe subscribes to an `Observable` or `Promise` and returns the latest value it has
 * emitted. When a new value is emitted, the `push` pipe will run change detection and it works
 * even when `zone` has been disabled. When the component gets destroyed,
 * the `push` pipe unsubscribes automatically to avoid potential memory leaks.
 *
 */
var PushPipe = /** @class */ (function () {
    function PushPipe(_ref) {
        this._ref = _ref;
        this._latestValue = null;
        this._latestReturnedValue = null;
        this._subscription = null;
        this._obj = null;
        this._strategy = (/** @type {?} */ (null));
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    PushPipe.prototype.transform = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        if (this._obj === null) {
            if (obj != null) {
                this._subscribe(obj);
            }
            this._latestReturnedValue = this._latestValue;
            return this._latestValue;
        }
        if (obj !== this._obj) {
            this._dispose();
            return this.transform((/** @type {?} */ (obj)));
        }
        if (this._latestValue === this._latestReturnedValue) {
            return this._latestReturnedValue;
        }
        this._latestReturnedValue = this._latestValue;
        return WrappedValue.wrap(this._latestValue);
    };
    /**
     * @return {?}
     */
    PushPipe.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._subscription !== null) {
            this._dispose();
        }
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    PushPipe.prototype._subscribe = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        this._obj = obj;
        this._strategy = this._selectStrategy(obj);
        this._subscription = this._strategy.createSubscription(obj, (/**
         * @param {?} value
         * @return {?}
         */
        function (value) { return _this._updateLatestValue(obj, value); }));
    };
    /**
     * @private
     * @return {?}
     */
    PushPipe.prototype._dispose = /**
     * @private
     * @return {?}
     */
    function () {
        this._strategy.dispose((/** @type {?} */ (this._subscription)));
        this._latestValue = null;
        this._latestReturnedValue = null;
        this._subscription = null;
        this._obj = null;
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    PushPipe.prototype._selectStrategy = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        if (isPromise(obj)) {
            return _promiseStrategy;
        }
        if (isObservable(obj)) {
            return _observableStrategy;
        }
        throw invalidPipeArgumentError(PushPipe, obj);
    };
    /**
     * @private
     * @param {?} async
     * @param {?} value
     * @return {?}
     */
    PushPipe.prototype._updateLatestValue = /**
     * @private
     * @param {?} async
     * @param {?} value
     * @return {?}
     */
    function (async, value) {
        if (async === this._obj) {
            this._latestValue = value;
            this._ref.detectChanges();
        }
    };
    PushPipe.decorators = [
        { type: Pipe, args: [{ name: 'push', pure: false },] }
    ];
    /** @nocollapse */
    PushPipe.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    return PushPipe;
}());
export { PushPipe };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._latestValue;
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._latestReturnedValue;
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._subscription;
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._obj;
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._strategy;
    /**
     * @type {?}
     * @private
     */
    PushPipe.prototype._ref;
}
var PushPipeModule = /** @class */ (function () {
    function PushPipeModule() {
    }
    PushPipeModule.decorators = [
        { type: NgModule, args: [{
                    exports: [PushPipe],
                    declarations: [PushPipe],
                },] }
    ];
    return PushPipeModule;
}());
export { PushPipeModule };
